sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/BatchRecorderSdk.java

```diff
@@ -10,61 +10,105 @@
 import io.opentelemetry.api.metrics.DoubleCounter;
 import io.opentelemetry.api.metrics.DoubleUpDownCounter;
 import io.opentelemetry.api.metrics.DoubleValueRecorder;
+import io.opentelemetry.api.metrics.Instrument;
 import io.opentelemetry.api.metrics.LongCounter;
 import io.opentelemetry.api.metrics.LongUpDownCounter;
 import io.opentelemetry.api.metrics.LongValueRecorder;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 /**
  * Minimal implementation of the {@link BatchRecorder} that simply redirects the calls to the
  * instruments.
- *
- * <p>TODO: Add an async queue processing to process batch records.
  */
 final class BatchRecorderSdk implements BatchRecorder {
   private final Labels labelSet;
 
+  private volatile Map<Instrument, List<Number>> recordings = new HashMap<>();
+  private volatile Map<Instrument, List<Number>> previousRecordings = new HashMap<>();
+
   BatchRecorderSdk(String... keyValuePairs) {
     this.labelSet = Labels.of(keyValuePairs);
   }
 
   @Override
   public BatchRecorder put(LongValueRecorder valueRecorder, long value) {
-    valueRecorder.record(value, labelSet);
+    recordings.computeIfAbsent(valueRecorder, instrument -> new ArrayList<>()).add(value);
     return this;
   }
 
   @Override
   public BatchRecorder put(DoubleValueRecorder valueRecorder, double value) {
-    valueRecorder.record(value, labelSet);
+    recordings.computeIfAbsent(valueRecorder, instrument -> new ArrayList<>()).add(value);
     return this;
   }
 
   @Override
   public BatchRecorder put(LongCounter counter, long value) {
-    counter.add(value, labelSet);
+    recordings.computeIfAbsent(counter, instrument -> new ArrayList<>()).add(value);
     return this;
   }
 
   @Override
   public BatchRecorder put(DoubleCounter counter, double value) {
-    counter.add(value, labelSet);
+    recordings.computeIfAbsent(counter, instrument -> new ArrayList<>()).add(value);
     return this;
   }
 
   @Override
   public BatchRecorder put(LongUpDownCounter upDownCounter, long value) {
-    upDownCounter.add(value, labelSet);
+    recordings.computeIfAbsent(upDownCounter, instrument -> new ArrayList<>()).add(value);
     return this;
   }
 
   @Override
   public BatchRecorder put(DoubleUpDownCounter upDownCounter, double value) {
-    upDownCounter.add(value, labelSet);
+    recordings.computeIfAbsent(upDownCounter, instrument -> new ArrayList<>()).add(value);
     return this;
   }
 
   @Override
   public void record() {
-    // No-op in this minimal implementation.
+    previousRecordings.clear();
+
+    Map<Instrument, List<Number>> temp = recordings;
+    recordings = previousRecordings;
+    previousRecordings = temp;
+
+    previousRecordings.forEach(
+        (instrument, numbers) -> {
+          if (instrument instanceof DoubleUpDownCounter) {
```

Comment:
If I'm reading this right, I think there's less code duplication if we have the loop on the outside of the instanceof checks

Reply:
another thing I went back and forth on a couple times. I doubt it makes much difference performance-wise, so if you think it's clearer the other way, happy to change it.

Reply:
done